wallet-new/src/Cardano/Wallet/API/V1/Errors.hs

```diff
@@ -1,19 +1,104 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
 
 module Cardano.Wallet.API.V1.Errors where
 
 import           Universum
 
-import           Cardano.Wallet.API.V1.Types (WalletError (..))
-import           Data.Aeson (encode)
+import           Data.Aeson
+import           Data.Aeson.TH (deriveJSON)
 import qualified Network.HTTP.Types.Header as HTTP
 import           Servant
+import           Test.QuickCheck (Arbitrary (..), oneof)
 
+import           Cardano.Wallet.API.V1.TH (conNamesList, deriveWalletErrorJSON)
+
+--
+-- Error handling
+--
+
+-- | Type representing any error which might be thrown by wallet.
+--
+-- Errors are represented in JSON in the following way:
+-- ```
+-- {
+--     "message" : <constr_name>,
+--     "diagnostic" : <data>
+-- }
+-- ```
+-- where `<constr_name>` is a string containing name of error's constructor (e. g. `NotEnoughMoney`),
+-- and `<data>` is an object containing additional error data.
+-- Additional data contains constructor fields, field names are record field names without
+-- a `we` prefix, e. g. for `OutputIsRedeem` error "diagnostic" field will be the following:
+-- ```
+-- {
+--     "address" : <address>
+-- }
+-- ```
+-- If constructor does not define record field names, then data values are labeled with
+-- keys "v1", "v2" and so on, e. g. for `NotRecordError` error:
+-- ```
+-- {
+--     "v1" : 3,
+--     "v2" : "blah"
+-- }
+-- ```
+-- If constructor does not have additional data (like in case of `WalletNotFound` error),
+-- then "diagnostic" field will be empty object.
+--
+-- TODO: change fields' types to actual Cardano core types, like `Coin` and `Address`
+data WalletError =
+      NotEnoughMoney { weNeedMore :: !Int }
+    | OutputIsRedeem { weAddress :: !Text }
+    | SomeOtherError { weFoo :: !Text, weBar :: !Int }
+    | NotRecordError !Int !Text
+    | MigrationFailed { weDescription :: !Text }
+    | WalletNotFound
```

Comment:
The only thing which makes me sad of this approach is that now all our field accessors are partial, i.e. `weDescription` has type `weDescription :: Text -> WalletError` but it will blow up if we use it to update anything different from a `MigrationFailed`.

Reply:
I don't think that updating already created `WalletError`s will be a common use case - we'll just create and immediately throw them, I suppose.
Even if such use case will occur (which I doubt), I cannot imagine why would a developer update `weDescription` field of given error without making sure that this particular error has this `weDescription`field. It feels similar to leaving `undefined` somewhere in the code - if you do this, you either know very well what you're doing or you're making a very silly mistake which cannot really be prevented by a compiler.

Reply:
Yea, I guess it shouldn't be the end of the world. As you said, the typical use case is to create a `WalletError`, not modifying it.